pkg/reconciler/route/reconcile_resources.go

```diff
@@ -123,37 +126,101 @@ func (c *Reconciler) reconcileClusterIngress(
 	return clusterIngress, err
 }
 
-func (c *Reconciler) reconcilePlaceholderService(ctx context.Context, route *v1alpha1.Route, ingress *netv1alpha1.ClusterIngress) error {
+func (c *Reconciler) deleteServices(namespace string, serviceNames sets.String) error {
+	for _, serviceName := range serviceNames.List() {
+		if err := c.KubeClientSet.CoreV1().Services(namespace).Delete(serviceName, nil); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Reconciler) getServiceNameSet(route *v1alpha1.Route) (sets.String, error) {
+	currentServices, err := c.serviceLister.List(resources.SelectorFromRoute(route))
+	if err != nil {
+		return nil, err
+	}
+
+	serviceNames := sets.NewString()
+	for _, svc := range currentServices {
+		serviceNames.Insert(svc.Name)
+	}
+
+	return serviceNames, nil
+}
+
+func (c *Reconciler) reconcilePlaceholderServices(ctx context.Context, route *v1alpha1.Route, targets map[string]traffic.RevisionTargets) ([]*corev1.Service, error) {
 	logger := logging.FromContext(ctx)
 	ns := route.Namespace
-	name := resourcenames.K8sService(route)
 
-	desiredService, err := resources.MakeK8sService(route, ingress)
+	names := make([]string, 0, len(targets))
```

Comment:
can we have `names` be a `sets.String` ?

Reply:
and done!